File: backend/app/services/message_service.py
from sqlalchemy.orm import Session
from sqlalchemy import desc
from fastapi import HTTPException, status
from typing import List
from app.models.message import Message, UserDeletedMessage
from app.models.conversation import Conversation
from app.models.user import User
from app.schemas import MessageCreate, MessageResponse
from app.utils.encryption import message_encryption
from app.config import get_settings

settings = get_settings()


class MessageService:
    """Service for message operations."""
    
    @staticmethod
    def create_message(
        db: Session,
        message_data: MessageCreate,
        sender: User
    ) -> MessageResponse:
        """Create a new message with encryption."""
        # Verify conversation exists and user is participant
        conversation = db.query(Conversation).filter(
            Conversation.id == message_data.conversation_id
        ).first()
        
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Conversation not found"
            )
        
        if sender not in conversation.participants:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not a participant in this conversation"
            )
        
        # Encrypt message content
        encrypted_content = message_encryption.encrypt(message_data.content)
        
        # Create message
        new_message = Message(
            content=encrypted_content,
            conversation_id=message_data.conversation_id,
            sender_id=sender.id,
            file_path=message_data.file_path,
            file_type=message_data.file_type
        )
        
        db.add(new_message)
        db.commit()
        db.refresh(new_message)
        
        # Prune old messages if limit exceeded
        MessageService._prune_old_messages(db, message_data.conversation_id)
        
        # Return decrypted message
        return MessageService._message_to_response(new_message)
    
    @staticmethod
    def get_conversation_messages(
        db: Session,
        conversation_id: int,
        user: User,
        limit: int = None
    ) -> List[MessageResponse]:
        """Get messages for a conversation (last N messages)."""
        # Verify conversation exists and user is participant
        conversation = db.query(Conversation).filter(
            Conversation.id == conversation_id
        ).first()
        
        if not conversation:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Conversation not found"
            )
        
        if user not in conversation.participants:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="You are not a participant in this conversation"
            )
        
        # Get last N messages
        if limit is None:
            limit = settings.MAX_MESSAGES_PER_CONVERSATION
        
        # Subquery to filter out deleted messages for the user
        deleted_for_me = db.query(UserDeletedMessage.message_id).filter(
            UserDeletedMessage.user_id == user.id
        )

        messages = db.query(Message).filter(
            Message.conversation_id == conversation_id,
            ~Message.id.in_(deleted_for_me)
        ).order_by(desc(Message.created_at)).limit(limit).all()
        
        # Reverse to chronological order and decrypt
        messages.reverse()
        return [MessageService._message_to_response(msg) for msg in messages]

    @staticmethod
    def edit_message(db: Session, message_id: int, user_id: int, content: str) -> MessageResponse:
        """Edit a message content."""
        message = db.query(Message).filter(Message.id == message_id).first()
        if not message:
            raise HTTPException(status_code=404, detail="Message not found")
            
        if message.sender_id != user_id:
            raise HTTPException(status_code=403, detail="Not authorized to edit this message")
            
        if message.is_deleted:
            raise HTTPException(status_code=400, detail="Cannot edit a deleted message")
            
        # Encrypt updated content
        encrypted_content = message_encryption.encrypt(content)
        message.content = encrypted_content
        message.is_edited = True
        
        db.commit()
        db.refresh(message)
        return MessageService._message_to_response(message)

    @staticmethod
    def delete_message(db: Session, message_id: int, user_id: int, delete_type: str) -> MessageResponse:
        """Delete a message (for me or for everyone)."""
        message = db.query(Message).filter(Message.id == message_id).first()
        if not message:
            raise HTTPException(status_code=404, detail="Message not found")
            
        if delete_type == "for_everyone":
            if message.sender_id != user_id:
                raise HTTPException(status_code=403, detail="Not authorized to delete this message for everyone")
            
            message.is_deleted = True
            db.commit()
            return MessageService._message_to_response(message)
            
        elif delete_type == "for_me":
             # Check if already deleted for this user
            existing = db.query(UserDeletedMessage).filter_by(
                user_id=user_id, 
                message_id=message_id
            ).first()
            
            if not existing:
                user_deleted = UserDeletedMessage(user_id=user_id, message_id=message_id)
                db.add(user_deleted)
                db.commit()
            
            # Since message is deleted for user, return None or handled by caller?
            # Returns message but caller should know it's deleted for them.
            # But the websocket event should probably not be broadcast if it's "for_me"
            return MessageService._message_to_response(message)
            
        else:
             raise HTTPException(status_code=400, detail="Invalid delete type")
    
    @staticmethod
    def _prune_old_messages(db: Session, conversation_id: int):
        """Remove messages beyond the limit."""
        message_count = db.query(Message).filter(
            Message.conversation_id == conversation_id
        ).count()
        
        if message_count > settings.MAX_MESSAGES_PER_CONVERSATION:
            # Get IDs of messages to delete
            messages_to_delete = db.query(Message.id).filter(
                Message.conversation_id == conversation_id
            ).order_by(Message.created_at).limit(
                message_count - settings.MAX_MESSAGES_PER_CONVERSATION
            ).all()
            
            # Delete old messages
            db.query(Message).filter(
                Message.id.in_([m.id for m in messages_to_delete])
            ).delete(synchronize_session=False)
            
            db.commit()
    
    @staticmethod
    def _message_to_response(message: Message) -> MessageResponse:
        """Convert Message model to response with decryption."""
        decrypted_content = message_encryption.decrypt(message.content)
        
        return MessageResponse(
            id=message.id,
            content=decrypted_content,
            conversation_id=message.conversation_id,
            sender_id=message.sender_id,
            sender_username=message.sender.username,
            created_at=message.created_at,
            file_path=message.file_path,
            file_type=message.file_type,
            is_deleted=bool(message.is_deleted),
            is_edited=bool(message.is_edited)
        )

File: backend_go/internal/service/message_service.go
package service

import (
	"context"
	"errors"
	"fmt"
	"time"

	"backend_go/internal/domain"
	"backend_go/internal/security"
)

// Sentinel errors used by handlers to map to HTTP status codes.
var (
	ErrForbidden      = errors.New("forbidden")
	ErrMessageDeleted = errors.New("message is already deleted")
)

type MessageService struct {
	conversations domain.ConversationRepository
	participants  domain.ParticipantRepository
	messages      domain.MessageRepository
	deletedMsgs   domain.UserDeletedMessageRepository
	users         domain.UserRepository
	encryptor     *security.Encryptor

	MaxMessagesPerConversation int
}

func NewMessageService(
	conversations domain.ConversationRepository,
	participants domain.ParticipantRepository,
	messages domain.MessageRepository,
	deletedMsgs domain.UserDeletedMessageRepository,
	users domain.UserRepository,
	encryptor *security.Encryptor,
	maxMessages int,
) *MessageService {
	return &MessageService{
		conversations:              conversations,
		participants:               participants,
		messages:                   messages,
		deletedMsgs:                deletedMsgs,
		users:                      users,
		encryptor:                  encryptor,
		MaxMessagesPerConversation: maxMessages,
	}
}

type MessageCreateInput struct {
	ConversationID int64
	Content        string
	FilePath       *string
	FileType       *string
}

func (s *MessageService) CreateMessage(
	ctx context.Context,
	in MessageCreateInput,
	senderID int64,
) (*domain.Message, error) {
	if len([]rune(in.Content)) > 5000 {
		return nil, errors.New("message content exceeds 5000 characters")
	}

	conv, err := s.conversations.GetByID(ctx, in.ConversationID)
	if err != nil {
		return nil, fmt.Errorf("get conversation: %w", err)
	}
	if conv == nil {
		return nil, errors.New("conversation not found")
	}
	isParticipant, err := s.participants.IsParticipant(ctx, in.ConversationID, senderID)
	if err != nil {
		return nil, fmt.Errorf("check participant: %w", err)
	}
	if !isParticipant {
		return nil, errors.New("you are not a participant in this conversation")
	}

	if in.Content == "" && (in.FilePath == nil || *in.FilePath == "") {
		return nil, errors.New("message content cannot be empty")
	}

	encrypted, err := s.encryptor.Encrypt(in.Content)
	if err != nil {
		return nil, fmt.Errorf("encrypt content: %w", err)
	}

	msg := &domain.Message{
		Content:        encrypted,
		ConversationID: in.ConversationID,
		SenderID:       senderID,
		FilePath:       in.FilePath,
		FileType:       in.FileType,
		IsDeleted:      false,
	}

	if err := s.messages.Create(ctx, msg); err != nil {
		return nil, err
	}

	if s.MaxMessagesPerConversation > 0 {
		if err := s.messages.PruneOld(ctx, in.ConversationID, s.MaxMessagesPerConversation); err != nil {
			return nil, fmt.Errorf("prune old messages: %w", err)
		}
	}

	return msg, nil
}

func (s *MessageService) EditMessage(
	ctx context.Context,
	callerID, messageID int64,
	newContent string,
) (*domain.Message, error) {
	if len([]rune(newContent)) > 5000 {
		return nil, errors.New("message content exceeds 5000 characters")
	}

	msg, err := s.messages.GetByID(ctx, messageID)
	if err != nil {
		return nil, fmt.Errorf("get message: %w", err)
	}
	if msg == nil {
		return nil, errors.New("message not found")
	}
	if msg.IsDeleted {
		return nil, ErrMessageDeleted
	}
	if msg.SenderID != callerID {
		return nil, ErrForbidden
	}

	encrypted, err := s.encryptor.Encrypt(newContent)
	if err != nil {
		return nil, fmt.Errorf("encrypt content: %w", err)
	}

	msg.Content = encrypted
	msg.IsEdited = true
	if err := s.messages.Update(ctx, msg); err != nil {
		return nil, fmt.Errorf("update message: %w", err)
	}
	return msg, nil
}

func (s *MessageService) DeleteMessage(
	ctx context.Context,
	callerID, messageID int64,
	deleteType string, // "for_me" | "for_everyone"
) (*domain.Message, error) {
	msg, err := s.messages.GetByID(ctx, messageID)
	if err != nil {
		return nil, fmt.Errorf("get message: %w", err)
	}
	if msg == nil {
		return nil, errors.New("message not found")
	}

	switch deleteType {
	case "for_everyone":
		if msg.SenderID != callerID {
			return nil, ErrForbidden
		}
		if err := s.messages.SoftDeleteForEveryone(ctx, messageID); err != nil {
			return nil, fmt.Errorf("soft delete: %w", err)
		}
		msg.IsDeleted = true
	case "for_me":
		if err := s.deletedMsgs.Create(ctx, callerID, messageID); err != nil {
			return nil, fmt.Errorf("delete for me: %w", err)
		}
	default:
		return nil, errors.New("delete_type must be 'for_me' or 'for_everyone'")
	}

	return msg, nil
}

func (s *MessageService) ListMessages(
	ctx context.Context,
	conversationID int64,
	userID int64,
	limit int,
) ([]*domain.Message, error) {
	conv, err := s.conversations.GetByID(ctx, conversationID)
	if err != nil {
		return nil, fmt.Errorf("get conversation: %w", err)
	}
	if conv == nil {
		return nil, errors.New("conversation not found")
	}
	isParticipant, err := s.participants.IsParticipant(ctx, conversationID, userID)
	if err != nil {
		return nil, fmt.Errorf("check participant: %w", err)
	}
	if !isParticipant {
		return nil, errors.New("you are not a participant in this conversation")
	}

	if limit <= 0 || limit > s.MaxMessagesPerConversation {
		limit = s.MaxMessagesPerConversation
	}

	msgs, err := s.messages.ListForConversationForUser(ctx, conversationID, userID, limit)
	if err != nil {
		return nil, err
	}

	// Reverse to chronological order (DB returns DESC)
	for i, j := 0, len(msgs)-1; i < j; i, j = i+1, j-1 {
		msgs[i], msgs[j] = msgs[j], msgs[i]
	}
	return msgs, nil
}

func (s *MessageService) MarkAllReadInConversation(ctx context.Context, conversationID, callerID int64) error {
	isParticipant, err := s.participants.IsParticipant(ctx, conversationID, callerID)
	if err != nil {
		return fmt.Errorf("check participant: %w", err)
	}
	if !isParticipant {
		return ErrForbidden
	}

	return s.messages.MarkAllReadInConversation(ctx, conversationID, callerID)
}

// GetParticipantIDs returns user IDs of all conversation participants (for WS broadcasts).
func (s *MessageService) GetParticipantIDs(ctx context.Context, conversationID int64) ([]int64, error) {
	participants, err := s.participants.ListParticipants(ctx, conversationID)
	if err != nil {
		return nil, err
	}
	ids := make([]int64, len(participants))
	for i, p := range participants {
		ids[i] = p.ID
	}
	return ids, nil
}

// MessageResponse mirrors the API response expected by the frontend.
type MessageResponse struct {
	ID             int64     `json:"id"`
	Content        string    `json:"content"`
	ConversationID int64     `json:"conversation_id"`
	SenderID       int64     `json:"sender_id"`
	SenderUsername string    `json:"sender_username"`
	CreatedAt      time.Time `json:"created_at"`
	FilePath       *string   `json:"file_path,omitempty"`
	FileType       *string   `json:"file_type,omitempty"`
	IsDeleted      bool      `json:"is_deleted"`
	IsEdited       bool      `json:"is_edited"`
	IsRead         bool      `json:"is_read"`
}

// ToResponse converts a domain message into a decrypted response DTO.
func (s *MessageService) ToResponse(ctx context.Context, m *domain.Message) (*MessageResponse, error) {
	content := m.Content
	if !m.IsDeleted {
		dec, err := s.encryptor.Decrypt(m.Content)
		if err == nil {
			content = dec
		}
		// on decrypt error fall back to raw (mirrors Python behaviour)
	}
	var username string
	if u, err := s.users.GetByID(ctx, m.SenderID); err == nil && u != nil {
		username = u.Username
	}
	return &MessageResponse{
		ID:             m.ID,
		Content:        content,
		ConversationID: m.ConversationID,
		SenderID:       m.SenderID,
		SenderUsername: username,
		CreatedAt:      m.CreatedAt,
		FilePath:       m.FilePath,
		FileType:       m.FileType,
		IsDeleted:      m.IsDeleted,
		IsEdited:       m.IsEdited,
		IsRead:         m.IsRead,
	}, nil
}

// ToResponses converts a slice of domain messages into response DTOs.
func (s *MessageService) ToResponses(ctx context.Context, msgs []*domain.Message) ([]*MessageResponse, error) {
	res := make([]*MessageResponse, 0, len(msgs))
	for _, m := range msgs {
		dto, err := s.ToResponse(ctx, m)
		if err != nil {
			return nil, err
		}
		res = append(res, dto)
	}
	return res, nil
}
